src/main/java/com/minecolonies/core/tileentities/TileEntityColonyBuilding.java

```diff
@@ -595,6 +597,18 @@ public IItemHandler getItemHandlerCap(final Direction side)
                                 {
                                     handlers.add(rack.getInventory());
                                     rack.setBuildingPos(this.getBlockPos());
+                                    if (world instanceof ServerLevel serverLevel)
+                                    {
+                                        serverLevel.registerCapabilityListener(rack.getBlockPos(), () -> {
+                                            if (!WorldUtil.isBlockLoaded(world, pos) || world.getBlockEntity(pos) != te)
+                                            {
+                                                return false;
+                                            }
+                                            this.combinedInv = null;
+                                            this.invalidateCapabilities();
+                                            return true;
+                                        });
+                                    }
                                 }
                                 else
                                 {
```

Comment:
I'd probably move the else block to the `tick` method, continuously reading each container position to see if it's loaded and is a rack, removing if not, setting the building pos if it is.

Reply:
Removals should already be handled through the capability invalidation listener and the `getOrCreateCombinedInv` call in `tick`.

In the case where the player has broken a rack and later re-places it, it won't get picked up until a chunk reload, building upgrade, or removal of another rack. The goal of using the listeners was to avoid querying the block entities every tick, so maybe we should check for re-placed racks every X (100?) ticks?

Reply:
The thing is, I don't want to see another solution using yet another different system.
If this is not directly crucial to be removed asap, then we are better off moving this check to the building directly using the `onColonyTick` method. Which is a tick happening about every 30 seconds.

Reply:
Putting the check in `onColonyTick` seems reasonable then. But I think we should keep the capability listeners, they seem to be a tool built for this purpose. Without it there could be errors if something tries to query the inventory of a broken rack before the next colony tick.